lib/dyn.ml

```diff
@@ -11,6 +11,26 @@ sig
   type 'a t
 end
 
+module type MapS =
+sig
+  type t
+  type 'a u
+  type 'a tag_
+  val empty : t
```

Comment:
Nitpick: for coherence with Map.S, I'd rather replace `t` with `key` and `u` with `t`. Also, exporting `tag_` with an underscore seems utterly ugly to me, it'd be nicer to just call it `tag` even if that means we have to fiddle with the constraint below (although IIRC, constraints are not recursive?).

Reply:
OK. For `tag_`, I had a conflict at the time of writing `with type 'a tag = 'a tag`. This is why I fell back on `tag_` in `dyn.ml`. Maybe a `non rec` would work, I did not try. What are the usages in this kind of situation?

Reply:
`type nonrec` isn't used in `with` but you need it for line 111 below.
In other words this toy example works
```ocaml

module Bar : sig type t val v : t end = struct
  type t = int
  let v = 0
end
open Bar

module rec Foo : sig type t val v : t -> t end with type t = t = struct
  type nonrec t = t
    let v x = x
end

let _ = Foo.v Bar.v
```

Reply:
Actually, I'm confused. E.g., the type of `Map.add` is `key -> 'a -> 'a t -> 'a t`. Then, don't you rather mean `'a tag_` -> `'a key` (and something else for `'a u`)?

Reply:
Indeed, I mixed things up. So in the signature, I'd see
```
type t
type 'a obj
type 'a key
```
and the functor constraints this to
```
type 'a key = 'a tag
type 'a obj = 'a M.t
```